Generic hashing for the new closure representation

xavierleroy opened the pull request:
This PR revises the generic hash functions to take advantage of the new closure representation introduced in #9619: code pointers are directly mixed into the hash rather than having to be detected using `Is_in_value_area`.

As a consequence, the `Is_in_value_area` test can be removed altogether in no-naked-pointers mode, which should make hashing faster.

Currently the new code for closures is activated only in no-naked-pointers mode, even though it is sound in naked-pointers mode too.

Closes: #2168

xavierleroy committed:
Generic hashing for the new closure representation

This commit revises the generic hash functions to take advantage of
the new closure representation: code pointers are directly mixed
into the hash rather than having to be detected using Is_in_value_area.

Currently the new code for closures is activated only in no-naked-pointers
mode, even though it is sound in naked-pointers mode too.

Closes: #2168

stedolan commented:
Since closures are supported both in marshalling and in hashing, it would be nice if marshalling and unmarshalling a closure did not change its hash.

This would have been painful to achieve before, but I think it can now be done easily enough by hashing `codeptr - cf->code_start` instead of `codeptr`, after looking up `cf` using the approach of #9649?

xavierleroy commented:
> hashing codeptr - cf->code_start instead of codeptr

This can be done, and with #9654 it might even be not too slow :-)

For additional fun, we could even mix in (some bits of) the digest for the code fragment.  But this may be going too far.

One caveat, though: I'd like to have the same hash values in naked-pointers mode and in no-naked-pointers mode.  So, this means activating the new hashing schema for closures in both modes, and not just in no-naked-pointers mode, as done conservatively in this PR.

stedolan commented:
> One caveat, though: I'd like to have the same hash values in naked-pointers mode and in no-naked-pointers mode. So, this means activating the new hashing schema for closures in both modes, and not just in no-naked-pointers mode, as done conservatively in this PR.

That would indeed be nice, but even having deterministic hash values in either of these modes would be an improvement. Currently, and even with this PR, closures do not have the same hash values between two runs of the same binary. (At least on PIE systems with ASLR, or in the presence of dynlink)

xavierleroy commented:
OK, I'll implement the "relative code pointer" approach.  Since #9654 changes the handling of code fragments, I'd rather wait until #9654 is reviewed and merged.

xavierleroy committed:
Generic hashing: use relative offsets for code pointers

As suggested by @stedolan during review of #9649.

For a code pointer, instead of hashing its absolute address,
hash the relative offset of the code pointer within its code fragment.

This way, the hash is reproducible between runs of the program,
even if absolute addresses are not reproducible (as with ASLR).

xavierleroy commented:
Rebased on trunk and implemented @stedolan's "relative code pointer" approach.  This is ready for review.

xavierleroy commented:
This needs to be reviewed.  @stedolan ? @damiendoligez ? Please ?

stedolan commented:
Did this get merged?

xavierleroy commented:
Oups, I made a mistake...

xavierleroy commented:
I can't resurrect this PR, so I made a new one at #9689.  Apologies for my clumsiness.
